File: problems/2376_count-special-integers/solution.go
import "strconv"

func countSpecialNumbers(n int64) int {
	perm := func(a int64, k int64) int64 {
		p := int64(1)
		for i := int64(0); i < k; i++ {
			p *= a - i
		}
		return p
	}

	s := strconv.FormatInt(n, 10)
	L := len(s)
	total := int64(0)
	for k := int64(1); k < int64(L); k++ {
		total += 9 * perm(9, k-1)
	}
	used := 0
	broke := false
	for i := 0; i < L; i++ {
		d := int(s[i] - '0')
		lo := 0
		if i == 0 {
			lo = 1
		}
		for x := lo; x < d; x++ {
			if used>>uint(x)&1 == 0 {
				total += perm(int64(10-(i+1)), int64(L-i-1))
			}
		}
		if used>>uint(d)&1 == 1 {
			broke = true
			break
		}
		used |= 1 << uint(d)
	}
	if !broke {
		total++
	}
	return int(total)
}

File: problems/2376_count-special-integers/solution.ts
function countSpecialNumbers(n: number): number {
    function perm(a: number, k: number): number {
        let p = 1;
        for (let i = 0; i < k; i++) {
            p *= a - i;
        }
        return p;
    }

    const digits = String(n).split("").map(Number);
    const L = digits.length;
    let total = 0;
    for (let k = 1; k < L; k++) {
        total += 9 * perm(9, k - 1);
    }
    let used = 0;
    let broke = false;
    for (let i = 0; i < L; i++) {
        const d = digits[i];
        for (let x = i === 0 ? 1 : 0; x < d; x++) {
            if (!((used >> x) & 1)) {
                total += perm(10 - (i + 1), L - i - 1);
            }
        }
        if ((used >> d) & 1) {
            broke = true;
            break;
        }
        used |= 1 << d;
    }
    if (!broke) {
        total += 1;
    }
    return total;
}
